undeclared identifier C error when using a function variable that is inherited in a closure in a callback argument

### Describe the bug

-

### Reproduction Steps

```v
pub type MyCallback = fn () | fn (ctx voidptr)

fn my_lower_level_func(func fn (ctx voidptr), ctx voidptr) {
	println('Bar')
}

fn my_func(cb MyCallback, ctx voidptr) {
	my_lower_level_func(fn [cb] (ctx voidptr) {
		match cb {
			fn () {
				cb()
			}
			fn (ctx voidptr) {
				cb(ctx)
			}
		}
		/* 
		// Workaround
		cb_ := cb
		match cb_ {
			fn () {
				cb_()
			}
			fn (ctx voidptr) {
				cb_(ctx)
			}
		}
		*/
	}, ctx)
}

my_func(fn () {
	println('Foo')
}, unsafe { nil })
```

### Expected Behavior

Works without additional declaration.

### Current Behavior

```
==================
/tmp/v_501/closure_cb.16995009286752534824.tmp.c:2290:16: error: use of undeclared identifier 'cb'
                (*(anon_fn_)(cb._anon_fn_))();
                             ^
/tmp/v_501/closure_cb.16995009286752534824.tmp.c:2293:23: error: use of undeclared identifier 'cb'
                (*(anon_fn_voidptr)(cb._anon_fn_voidptr))(ctx);
                                    ^
2 errors generated.
...
==================
```

### Possible Solution

_No response_

### Additional Information/Context

Repro is not as minimal as it could be. Please let me know if refining it would be of significant help.

### V version

-

### Environment details (OS name and version, etc.)

```
V full version: V 0.4.2 20a1af3.48a1d6c
OS: macos, macOS, 14.0, 23A344
Processor: 8 cpus, 64bit, little endian, Apple M1

getwd: /Users/t/Dev/v/playground
vexe: /Users/t/Dev/v/v/v
vexe mtime: 2023-10-21 12:14:56

vroot: OK, value: /Users/t/Dev/v/v
VMODULES: OK, value: /Users/t/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.42.0
Git vroot status: weekly.2023.42-19-g48a1d6cc
.git/config present: true

CC version: Apple clang version 14.0.3 (clang-1403.0.22.14.1)
thirdparty/tcc status: thirdparty-macos-arm64 a668e5a0-dirty
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/fn.v
+++ b/vlib/v/gen/c/fn.v
@@ -1777,7 +1777,11 @@ fn (mut g Gen) fn_call(node ast.CallExpr) {
 								cast_sym := g.table.sym(g.unwrap_generic(typ))
 								mut is_ptr := false
 								if i == 0 {
-									g.write(node.name)
+									if obj.is_inherited {
+										g.write(c.closure_ctx + '->' + node.name)
+									} else {
+										g.write(node.name)
+									}
 									if obj.orig_type.is_ptr() {
 										is_ptr = true
 									}

--- a/vlib/v/tests/closure_variable_in_smartcast_test.v
+++ b/vlib/v/tests/closure_variable_in_smartcast_test.v
@@ -1,0 +1,25 @@
+pub type MyCallback = fn () | fn (ctx voidptr)
+
+fn my_lower_level_func(func fn (ctx voidptr), ctx voidptr) {
+	println('Bar')
+}
+
+fn my_func(cb MyCallback, ctx voidptr) {
+	my_lower_level_func(fn [cb] (ctx voidptr) {
+		match cb {
+			fn () {
+				cb()
+			}
+			fn (ctx voidptr) {
+				cb(ctx)
+			}
+		}
+	}, ctx)
+}
+
+fn test_closure_variable_in_smartcast() {
+	my_func(fn () {
+		println('Foo')
+	}, unsafe { nil })
+	assert true
+}

